projectq/backends/_unitary.py

```diff
@@ -0,0 +1,191 @@
+# -*- coding: utf-8 -*-
+#   Copyright 2021 ProjectQ-Framework (www.projectq.ch)
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""
+Contains a backend that saves the unitary of the circuit
+"""
+
+from copy import deepcopy
+import itertools
+import math
+import numpy as np
+import warnings
+
+from projectq.cengines import BasicEngine
+from projectq.ops import (
+    AllocateQubitGate,
+    DeallocateQubitGate,
+    MeasureGate,
+    FlushGate,
+)
+
+
+def _qidmask(target_ids, control_ids, control_state, n_qubits):
+    """
+    Calculate index masks
+
+    Args:
+        target_ids (list): list of target qubit indices
+        control_ids (list): list of control qubit indices
+        control_state (list): list of states for the control qubits (0 or 1)
+        n_qubits (int): number of qubits
+    """
+    mask_list = []
+    perms = np.array([x[::-1] for x in itertools.product("01", repeat=n_qubits)]).astype(int)
+    all_ids = np.array(range(n_qubits))
+    irel_ids = np.delete(all_ids, control_ids + target_ids)
+    m = len(irel_ids)
+
+    if len(control_ids) > 0:
+        cstate = np.array(list(control_state)).astype(int)
+        cmask = np.where(np.all(perms[:, control_ids] == cstate, axis=1))
```

Comment:
I forgot to ask you, if we anyway do not support negatively controlled qubits (with `state == 0`) what is the point of using the `state` argument in the first place?

Reply:
It was an attempt to fix a bug but turns out that was not the cause: all incoming ctrl_state should be 111... due to decomp rule. This is removed.

Reply:
Ok.